scimath/Mathematics/ClassicalStatistics.tcc

```diff
@@ -940,78 +917,168 @@ vector<vector<uInt64> > ClassicalStatistics<CASA_STATP>::_binCounts(
         ++iDesc;
     }
     _initIterators();
+    uInt nThreadsMax = _nThreadsMax();
+    PtrHolder<vector<vector<uInt64> > > tBins(
+        new vector<vector<uInt64> >[CACHE_PADDING*nThreadsMax], True
+    );
+    PtrHolder<vector<CountedPtr<AccumType> > > tSameVal(
+        new vector<CountedPtr<AccumType> >[CACHE_PADDING*nThreadsMax], True
+    );
+    PtrHolder<vector<Bool> > tAllSame(
+        new vector<Bool>[CACHE_PADDING*nThreadsMax], True
+    );
+    for (uInt tid=0; tid<nThreadsMax; ++tid) {
+        uInt idx8 = CACHE_PADDING*tid;
+        tBins[idx8] = bins;
+        tSameVal[idx8] = sameVal;
+        tAllSame[idx8] = allSame;
+    }
     while (True) {
-        _initLoopVars();
-        if (_hasWeights) {
-            if (_hasMask) {
-                if (_hasRanges) {
-                    _findBins(
-                        bins, sameVal, allSame, _myData, _myWeights, _myCount,
-                        _myStride, _myMask, _maskStride, _myRanges, _myIsInclude,
-                        binDesc, maxLimit
-                    );
+        _initLoopVars(True);
+#pragma omp parallel for num_threads(_nthreads)
+        for (uInt i=0; i<_nBlocks; ++i) {
+#ifdef _OPENMP
+            uInt tid = omp_get_thread_num();
+#else
+            uInt tid = 0;
+#endif
+            uInt idx8 = CACHE_PADDING*tid;
+            uInt dataCount = _myCount - _offset[idx8] < BLOCK_SIZE ? _extra : BLOCK_SIZE;
+            _computeBins(
+                tBins[idx8], tSameVal[idx8], tAllSame[idx8], _dataIter[idx8],
+                _maskIter[idx8], _weightsIter[idx8], dataCount, binDesc, maxLimit
+            );
+            _incrementThreadIters(idx8);
+        }
+        if (_increment(False)) {
+            break;
+        }
+    }
+    vector<vector<uInt64> >::iterator begin = bins.begin();
+    vector<vector<uInt64> >::iterator iter = begin;
+    vector<vector<uInt64> >::iterator end = bins.end();
+    typename vector<CountedPtr<AccumType> >::iterator sbegin = sameVal.begin();
+    typename vector<CountedPtr<AccumType> >::iterator siter = sbegin;
+    typename vector<CountedPtr<AccumType> >::iterator send = sameVal.end();
+    vector<Bool>::iterator abegin = allSame.begin();
+    vector<Bool>::iterator aiter = abegin;
```

Comment:
these iterators do not need to be in the function global scope, vector.begin just maps to an inline &data[0] and this is not a performance relevant loop
also a short comment at the top on what you are doing helps readability. Its probably merging the thread results into the final results but I should not be required to read 40 lines of code to figure that out.